src/main/java/gregtech/api/util/InventoryUtils.java

```diff
@@ -90,29 +131,93 @@ public static boolean simulateItemStackMerge(List<ItemStack> items,
+        while(sourceItemStacks.hasNext())
+        {
             final ItemStack sourceItemStack = sourceItemStacks.next();
 
             // Find a matching item in the output bus, if any
             for(ItemStack destItemStack : destination)
-                if(ItemStack.areItemsEqual(destItemStack, sourceItemStack)) {
+                if(ItemStack.areItemsEqual(destItemStack, sourceItemStack) &&
+                   ItemStack.areItemStackTagsEqual(destItemStack, sourceItemStack))
+                {
                     // if it's possible to merge stacks
                     final int availableSlots = destItemStack.getMaxStackSize() - destItemStack.getCount();
-                    if(availableSlots > 0) {
+                    if(availableSlots > 0)
+                    {
                         final int itemCount = Math.min(availableSlots, sourceItemStack.getCount());
                         sourceItemStack.shrink(itemCount);
                         destItemStack.grow(itemCount);
 
                         // if the output stack was merged completely, remove it and stop looking
-                        if(sourceItemStack.isEmpty()) {
+                        if(sourceItemStack.isEmpty())
+                        {
                             sourceItemStacks.remove();
                             break;
                         }
                     }
                 }
         }
     }
+
+    /**
+     * Normalizes a potentially problematic ItemStack by splitting it into copied stacks of at most
+     * {@link ItemStack#getMaxStackSize()} items. Empty ItemStacks produce an empty
+     * result list, and stacks already in normalized form are copied as a singleton list.
+     *
+     * @param stack an ItemStack to process.
+     * @return an immutable List of the resulting ItemStacks, in descending size order.
+     */
+    public static List<ItemStack> normalizeItemStack(ItemStack stack)
+    {
+        if(stack.isEmpty())
+            return Collections.emptyList();
+
+        int maxCount = stack.getMaxStackSize();
+
+        if(stack.getCount() <= maxCount)
+            return Collections.singletonList(stack.copy());
+
+        return Collections.unmodifiableList(apportionStack(stack, maxCount));
+    }
+
+    /**
+     * Divides a stack of items into as many stacks of {@code maxCount} size as possible, followed by
+     * a partial stack of any remaining quantity. The original stack is not modified.
+     *
+     * @param stack    a non-empty ItemStack to split into stacks of at most {@code maxCount} items.
+     * @param maxCount the maximum number of items in each stack. Must be a positive, non-zero integer.
+     * @return the resulting zero to many whole stacks and zero to one partial stack.
+     */
+    public static List<ItemStack> apportionStack(ItemStack stack,
+                                                 final int maxCount)
+    {
+        assert !stack.isEmpty();
+        assert maxCount > 0;
```

Comment:
Please replace these with `if` check and `IllegalArgumentException`

Reply:
Addressed in 894b0cd